chemclipse/plugins/org.eclipse.chemclipse.msd.model/src/org/eclipse/chemclipse/msd/model/matrix/ExtractedMatrix.java

```diff
@@ -0,0 +1,158 @@
+/*******************************************************************************
+ * Copyright (c) 2020 Lablicate GmbH.
+ * 
+ * All rights reserved.
+ * This program and the accompanying materials are made available under the
+ * terms of the Eclipse Public License v1.0 which accompanies this distribution,
+ * and is available at http://www.eclipse.org/legal/epl-v10.html
+ * 
+ * Contributors:
+ * Lorenz Gerber - initial API and implementation
+ *******************************************************************************/
+package org.eclipse.chemclipse.msd.model.matrix;
+
+
+import java.util.Comparator;
+import java.util.Iterator;
+import java.util.List;
+import java.util.stream.Collectors;
+
+import org.eclipse.chemclipse.msd.model.core.AbstractIon;
+import org.eclipse.chemclipse.msd.model.core.IIon;
+import org.eclipse.chemclipse.msd.model.core.IIonBounds;
+import org.eclipse.chemclipse.msd.model.core.IScanMSD;
+import org.eclipse.chemclipse.msd.model.core.selection.IChromatogramSelectionMSD;
+import org.eclipse.chemclipse.msd.model.implementation.Ion;
+
+public class ExtractedMatrix {
+	
+	private IChromatogramSelectionMSD selection;
+	private List<IScanMSD> scans;
+	private int numberOfScans;
+	private int numberOfIons;
+	private int startIon;
+	private int stopIon;
+	private float[][] signal;
+	
+	public ExtractedMatrix(IChromatogramSelectionMSD chromatogramSelection) {
+		this.selection = chromatogramSelection;
+		this.numberOfScans = selection.getStopScan() - selection.getStartScan() + 1;
+		this.scans = extractScans();
+		if(checkHighRes( 10)) {
+			throw new IllegalArgumentException();
+		} else {
+			this.startIon = getMinMz();
+			this.stopIon = getMaxMz();
+			this.numberOfIons = this.stopIon - this.startIon + 1;
+			signal = new float[numberOfScans][this.numberOfIons];
+			List<IIon> currentIons;
+			int numberIonsCurrentScan;
+			for(int scanIndex = 0; scanIndex < numberOfScans; scanIndex++) {
+				currentIons = scans.get(scanIndex).getIons();
+				numberIonsCurrentScan = currentIons.size(); 
+				for(int j = 0; j < numberIonsCurrentScan; j++) {
+					signal[ scanIndex ] [((int) Math.round(currentIons.get(j).getIon() - this.startIon))] =  currentIons.get(j).getAbundance();
```

Comment:
You should extract 'currentIons.get(j)' to prevent additional look-up, from the coode it looks as if you even don't need the for-loop and could convert to enhancedforloop as the array index is calculated from the ion and not the running-index.